lib/spack/llnl/util/filesystem.py

```diff
@@ -902,19 +904,26 @@ def traverse_tree(source_root, dest_root, rel_path='', **kwargs):
         rel_child = os.path.join(rel_path, f)
 
         # Treat as a directory
-        # TODO: for symlinks, os.path.isdir looks for the link target. If the
-        # target is relative to the link, then that may not resolve properly
-        # relative to our cwd - see resolve_link_target_relative_to_the_link
-        if os.path.isdir(source_child) and (
-                follow_links or not os.path.islink(source_child)):
-
+        if os.path.isdir(source_child):
             # When follow_nonexisting isn't set, don't descend into dirs
             # in source that do not exist in dest
-            if follow_nonexisting or os.path.exists(dest_child):
-                tuples = traverse_tree(
-                    source_root, dest_root, rel_child, **kwargs)
-                for t in tuples:
-                    yield t
+            if not follow_nonexisting and not os.path.exists(dest_child):
+                continue
+
+            # When a symlink, ensure it's inside the prefix.
+            if os.path.islink(source_child):
+                if not follow_links:
+                    continue
```

Comment:
Perhaps we should have always done something analogous to doing `yield t` here (but don't traverse further); in that case, would you need to default to `follow_symlinks=True`?

Reply:
How do you mean exactly?

Reply:
Ah, I see what you mean. Now it yields a symlink to directory, and that is indeed required for copy_tree.

Reply:
@scheibelp does `ignore(path)` apply to (symlinks to) directories?

Reply:
It looks like that's what we do (from reading the code but we don't have tests which specifically enforce it (e.g. `test_symlinks_true_ignore` is the only test I see using `ignore`, but it doesn't try excluding symlinks). Are you seeing an issue with that?

Reply:
hm, just not sure what intended behavior is, but I'll leave it like this then